sdk/core/src/wallet/metamask.ts

```diff
@@ -0,0 +1,97 @@
+import { WalletBase } from './base.js';
+import {
+  BrowserProvider,
+  Transaction,
+  hexlify,
+  type TransactionRequest,
+  type Eip1193Provider,
+} from 'ethers';
+
+declare global {
+  // Expose the injected provider placed on the window object by MetaMask.
+  // We type it as unknown because the exact shape is library-specific and we
+  // interact with it exclusively through ethers' BrowserProvider wrapper.
+  interface Window {
+    ethereum?: unknown;
+  }
+}
+
+/**
+ * Wallet integration for the MetaMask browser extension.
+ *
+ * It fulfils the minimal `WalletBase` contract (fetching the current address
+ * and signing arbitrary messages). MetaMask **cannot** sign a raw transaction
+ * without also broadcasting it, therefore {@link signTxn} intentionally throws
```

Comment:
Does this apply to other wallets too? if so we might want to invert `WalletBase` then - have `sendTransaction` instead of `signTxn` and the `LocalWallet` can have the `signTxn` instead - so the `WalletBase` is completely usable

Reply:
I like this! Let me change it!